In [122]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

['titanic_test.csv', 'titanic_train.csv']


In [123]:
## reading the Input files. We can read any type of data(CSV,HTML,JSON etc)
train = pd.read_csv("../input/titanic_train.csv")
test = pd.read_csv("../input/titanic_test.csv")

In [124]:
## Just to get a feel of the train data. 
## head() displays first few rows whether tail() displays last few rows
train.head()

PassengerId  Survived  Pclass    ...        Fare Cabin  Embarked
0            1         0       3    ...      7.2500   NaN         S
1            2         1       1    ...     71.2833   C85         C
2            3         1       3    ...      7.9250   NaN         S
3            4         1       1    ...     53.1000  C123         S
4            5         0       3    ...      8.0500   NaN         S

[5 rows x 12 columns]

In [125]:
## Just to get a feel of the test data. 
test.head()

PassengerId  Pclass   ...    Cabin Embarked
0          892       3   ...      NaN        Q
1          893       3   ...      NaN        S
2          894       2   ...      NaN        Q
3          895       3   ...      NaN        S
4          896       3   ...      NaN        S

[5 rows x 11 columns]

In [126]:
## lets check what are the columns in Train.
print(train.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


In [127]:
## lets check what are the columns in Test.
print(test.columns)
## Survived column is missed. WHY?

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


In [128]:
## lets try to see the column details in TRAIN
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [129]:
## lets see the distribution of Target
train['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

In [130]:
## lets see some sample of Survived.
train[train['Survived'] == 1]

PassengerId  Survived  Pclass    ...         Fare Cabin  Embarked
1              2         1       1    ...      71.2833   C85         C
2              3         1       3    ...       7.9250   NaN         S
3              4         1       1    ...      53.1000  C123         S
8              9         1       3    ...      11.1333   NaN         S
9             10         1       2    ...      30.0708   NaN         C
10            11         1       3    ...      16.7000    G6         S
11            12         1       1    ...      26.5500  C103         S
15            16         1       2    ...      16.0000   NaN         S
17            18         1       2    ...      13.0000   NaN         S
19            20         1       3    ...       7.2250   NaN         C
21            22         1       2    ...      13.0000   D56         S
22            23         1       3    ...       8.0292   NaN         Q
23            24         1       1    ...      35.5000    A6         S
25            26         1       3    ...      31.3875   NaN         S
28            29         1       3    ...       7.8792   NaN         Q
31            32         1       1    ...     146.5208   B78         C
32            33         1       3    ...       7.7500   NaN         Q
36            37         1       3    ...       7.2292   NaN         C
39            40         1       3    ...      11.2417   NaN         C
43            44         1       2    ...      41.5792   NaN         C
44            45         1       3    ...       7.8792   NaN         Q
47            48         1       3    ...       7.7500   NaN         Q
52            53         1       1    ...      76.7292   D33         C
53            54         1       2    ...      26.0000   NaN         S
55            56         1       1    ...      35.5000   C52         S
56            57         1       2    ...      10.5000   NaN         S
58            59         1       2    ...      27.7500   NaN         S
61            62         1       1    ...      80.0000   B28       NaN
65            66         1       3    ...      15.2458   NaN         C
66            67         1       2    ...      10.5000   F33         S
..           ...       ...     ...    ...          ...   ...       ...
809          810         1       1    ...      53.1000    E8         S
820          821         1       1    ...      93.5000   B69         S
821          822         1       3    ...       8.6625   NaN         S
823          824         1       3    ...      12.4750  E121         S
827          828         1       2    ...      37.0042   NaN         C
828          829         1       3    ...       7.7500   NaN         Q
829          830         1       1    ...      80.0000   B28       NaN
830          831         1       3    ...      14.4542   NaN         C
831          832         1       2    ...      18.7500   NaN         S
835          836         1       1    ...      83.1583   E49         C
838          839         1       3    ...      56.4958   NaN         S
839          840         1       1    ...      29.7000   C47         C
842          843         1       1    ...      31.0000   NaN         C
849          850         1       1    ...      89.1042   C92         C
853          854         1       1    ...      39.4000   D28         S
855          856         1       3    ...       9.3500   NaN         S
856          857         1       1    ...     164.8667   NaN         S
857          858         1       1    ...      26.5500   E17         S
858          859         1       3    ...      19.2583   NaN         C
862          863         1       1    ...      25.9292   D17         S
865          866         1       2    ...      13.0000   NaN         S
866          867         1       2    ...      13.8583   NaN         C
869          870         1       3    ...      11.1333   NaN         S
871          872         1       1    ...      52.5542   D35         S
874          875         1   

## Some observations in Training set
* There are 891 rows.
* Age and Cabin looks like missing some data
* There are three kinds of data - int,float,Object

## lets see how the Age and Cabin looks like

## explore Age

In [131]:
train['Age'].head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

In [132]:
## lets see the statistics
train['Age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

In [133]:
## lets see how many missing values
train['Age'].isna().sum()

177

In [134]:
## lets try to replace missing AGE by average values
train['Age'] = train['Age'].fillna(train['Age'].mean())

In [135]:
## lets see how many missing values now.
train['Age'].isna().sum()

0

## explore Cabin

In [136]:
train['Cabin'].head()

0     NaN
1     C85
2     NaN
3    C123
4     NaN
Name: Cabin, dtype: object

In [137]:
## lets see the statistics
train['Cabin'].describe()

count     204
unique    147
top        G6
freq        4
Name: Cabin, dtype: object

In [138]:
## lets see how many missing values
train['Cabin'].isna().sum()

687

In [139]:
## lets see what are the unique values. 
train['Cabin'].unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [140]:
## lets replace NAN with character "Z" and see the NAN count
train['Cabin'] = train['Cabin'].fillna('Z')
train['Age'].isna().sum()
#train['Cabin'].str[0]

0

In [141]:
train['Cabin'] = train['Cabin'].str[0]

In [142]:
train['Cabin'].describe()

count     891
unique      9
top         Z
freq      687
Name: Cabin, dtype: object

Now see the Statistics. We have 9 different class representing the **Cabin** 

## Lets explore other variable to see how  they look

In [143]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [144]:
## lets see passenger ID
train['PassengerId'].unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

all unique values. Just to uniquly identify each passenger. 
Nothing to do with this variable during training.
We will drop this variable before training

train.drop(['PassengerId'],axis=1,inplace=True)

In [145]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [146]:
train['Survived'].unique()

array([0, 1])

This column defines the Predictions. Its either 0 or 1.

In [147]:
train['Pclass'].describe()

count    891.000000
mean       2.308642
std        0.836071
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64

In [148]:
train['Pclass'].unique()

array([3, 1, 2])

So ,there's 3 class representing different class of the people on boarded in  the Titanic. Its certainly a feature.

In [149]:
train['Name']

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
5                                       Moran, Mr. James
6                                McCarthy, Mr. Timothy J
7                         Palsson, Master. Gosta Leonard
8      Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                    Nasser, Mrs. Nicholas (Adele Achem)
10                       Sandstrom, Miss. Marguerite Rut
11                              Bonnell, Miss. Elizabeth
12                        Saundercock, Mr. William Henry
13                           Andersson, Mr. Anders Johan
14                  Vestrom, Miss. Hulda Amanda Adolfina
15                      Hewlett, Mrs. (Mary D Kingcome) 
16                                  Rice, Master. Eugene
17                          Wil

Can Name be a features ? Not Sure if it is a good feature. Lets drop it.

In [150]:
train.drop(['Name'],axis=1,inplace=True)

In [151]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [152]:
train['Sex'].unique() 

array(['male', 'female'], dtype=object)

In [153]:
train["Sex"].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [154]:
train['SibSp'].unique()

array([1, 0, 3, 4, 2, 5, 8])

In [155]:
train["SibSp"].value_counts()

0    608
1    209
2     28
4     18
3     16
8      7
5      5
Name: SibSp, dtype: int64

In [156]:
train['Parch'].unique()

array([0, 1, 2, 5, 3, 4, 6])

In [157]:
train["Parch"].value_counts()

0    678
1    118
2     80
5      5
3      5
4      4
6      1
Name: Parch, dtype: int64

In [158]:
train[['Ticket','Survived']]

Ticket  Survived
0           A/5 21171         0
1            PC 17599         1
2    STON/O2. 3101282         1
3              113803         1
4              373450         0
5              330877         0
6               17463         0
7              349909         0
8              347742         1
9              237736         1
10            PP 9549         1
11             113783         1
12          A/5. 2151         0
13             347082         0
14             350406         0
15             248706         1
16             382652         0
17             244373         1
18             345763         0
19               2649         1
20             239865         0
21             248698         1
22             330923         1
23             113788         1
24             349909         0
25             347077         1
26               2631         0
27              19950         0
28             330959         1
29             349216         0
..                ...       ...
861             28134         0
862             17466         1
863          CA. 2343         0
864            233866         0
865            236852         1
866     SC/PARIS 2149         1
867          PC 17590         0
868            345777         0
869            347742         1
870            349248         0
871             11751         1
872               695         0
873            345765         0
874         P/PP 3381         1
875              2667         1
876              7534         0
877            349212         0
878            349217         0
879             11767         1
880            230433         1
881            349257         0
882              7552         0
883  C.A./SOTON 34068         0
884   SOTON/OQ 392076         0
885            382652         0
886            211536         0
887            112053         1
888        W./C. 6607         0
889            111369         1
890            370376         0

[891 rows x 2 columns]

## tickets are unique. Can you find any features out of it. Please check

In [159]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [160]:
train['Fare']   ## Fare of the tickets. any relation?

0        7.2500
1       71.2833
2        7.9250
3       53.1000
4        8.0500
5        8.4583
6       51.8625
7       21.0750
8       11.1333
9       30.0708
10      16.7000
11      26.5500
12       8.0500
13      31.2750
14       7.8542
15      16.0000
16      29.1250
17      13.0000
18      18.0000
19       7.2250
20      26.0000
21      13.0000
22       8.0292
23      35.5000
24      21.0750
25      31.3875
26       7.2250
27     263.0000
28       7.8792
29       7.8958
         ...   
861     11.5000
862     25.9292
863     69.5500
864     13.0000
865     13.0000
866     13.8583
867     50.4958
868      9.5000
869     11.1333
870      7.8958
871     52.5542
872      5.0000
873      9.0000
874     24.0000
875      7.2250
876      9.8458
877      7.8958
878      7.8958
879     83.1583
880     26.0000
881      7.8958
882     10.5167
883     10.5000
884      7.0500
885     29.1250
886     13.0000
887     30.0000
888     23.4500
889     30.0000
890      7.7500
Name: Fare, Length: 891,

In [161]:
train['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [162]:
## Lets explore the test.csv. 